src/librbd/CreateRequest.cc

```diff
@@ -0,0 +1,633 @@
+    *result = cls_client::mirror_mode_get_finish(&it, &mirror_mode_internal);
+    if (*result < 0) {
+      lderr(cct) << "Failed to retrieve mirror mode" << dendl;
+      save_retval_and_cleanup(*result, &CreateRequest::remove_object_map);
+      return nullptr;
+    }
+  }
+
+  // TODO: remove redundant code...
+  switch (mirror_mode_internal) {
+  case cls::rbd::MIRROR_MODE_DISABLED:
+  case cls::rbd::MIRROR_MODE_IMAGE:
+  case cls::rbd::MIRROR_MODE_POOL:
+    mirror_mode = static_cast<rbd_mirror_mode_t>(mirror_mode_internal);
+    break;
+  default:
+    lderr(cct) << "Unknown mirror mode ("
+               << static_cast<uint32_t>(mirror_mode_internal) << ")"
+               << dendl;
+    save_retval_and_cleanup(-EINVAL, &CreateRequest::remove_object_map);
+    return nullptr;
+  }
+
+  journal_create();
+  return nullptr;
+}
+
+/*
+ * There doesn't exist native asynchronous call for journal creation.
+ * Executing this synchronously in rados callback context is prone to
+ * deadlocks
+ */
+void CreateRequest::journal_create() {
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  Context *ctx = new FunctionContext([this](int r) {
+      r = Journal<>::create(ioctx, id, journal_order, journal_splay_width,
+                            journal_pool, force_non_primary, primary_mirror_uuid);
+      handle_journal_create(r);
+    });
+
+  aio_work_queue->queue(ctx, 0);
+}
+
+void CreateRequest::handle_journal_create(int r) {
+  ldout(cct, 20) << __func__ << ": r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "error creating journal: " << cpp_strerror(r) << dendl;
+    save_retval_and_cleanup(r, &CreateRequest::remove_object_map);
+    return;
+  }
+
+  fetch_mirror_image();
+}
+
+void CreateRequest::fetch_mirror_image() {
+  if ((mirror_mode != RBD_MIRROR_MODE_POOL) && !force_non_primary) {
+    complete(0);
+    return;
+  }
+
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_image_get_start(&op, id);
+
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<CreateRequest, &CreateRequest::handle_fetch_mirror_image>(this);
+  outbl.clear();
+  int r = ioctx.aio_operate(RBD_MIRRORING, comp, &op, &outbl);
+  assert(r == 0);
+  comp->release();
+}
+
+Context *CreateRequest::handle_fetch_mirror_image(int *result) {
+  ldout(cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  // this should always be the case during create
+  assert(*result == -ENOENT);
+
+  mirror_image_enable();
+  return nullptr;
+}
+
+void CreateRequest::mirror_image_enable() {
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  mirror_image_internal.state = cls::rbd::MIRROR_IMAGE_STATE_ENABLED;
+  if (non_primary_global_image_id.empty()) {
+    uuid_d uuid_gen;
+    uuid_gen.generate_random();
+    mirror_image_internal.global_image_id = uuid_gen.to_string();
+  } else {
+    mirror_image_internal.global_image_id = non_primary_global_image_id;
```

Comment:
Note: this should be semi-temporary since I think rbd-mirror should do this before it attempts to create the image (w/ state of creating)